### PR TITLE
Add support for login via certificate

### PR DESCRIPTION
**Pull request by avandecreme**
The first commit is cherry picked from #62 

The second one is specific to this PR and allow one to use [certificate to login to vault](https://developer.hashicorp.com/vault/docs/auth/cert).

### CODE REVIEW

**Comment by Haennetz** [comment]
Hey thanks for your pull request. It would be nice if you can add a test for the certificate login. I merged #62 can you rebase your branch and remove the cherry-picked commit?

**Comment by avandecreme** [comment]
I added two commits.
The first one is implementing the tests. The second one is just to show how having https://github.com/jmgilman/dockertest-server/pull/15 merged could clean up this PR.

I committed the keys and certificates as well as the method to generate them. I see that gitguardian is not happy about it even though those are just test certificates.

I guess we could generate new certificates on every test run, provided we can run openssl in the runner. What do you think?

**Comment by Haennetz** [comment]
Yes I think generating a new certificate each time would be the best solution for that.
We can use [rcgen](https://docs.rs/rcgen/latest/rcgen/) as a dev dependency and create the certificate with it.

**Comment by avandecreme** [comment]
> Yes I think generating a new certificate each time would be the best solution for that. We can use [rcgen](https://docs.rs/rcgen/latest/rcgen/) as a dev dependency and create the certificate with it.

Done

**Comment by Haennetz** [comment]
Hey tanks for implementing this. The test is failing  at the moment because the nativ-tls feature didn't support pam only. It requires pkcs8 or pkcs12.

**Comment by avandecreme** [comment]
I added a commit to disable the test with native-tls since vaultrs does not support client certificates with it anyway. See https://github.com/jmgilman/vaultrs/blob/34e3874ee406723381ba99db9e5c77b64967294d/src/client.rs#L298

File: tests/cert.rs [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+use test_log::test;
+use vault_bind_mounts_container::{VaultServer, VaultServerConfig};
+use vaultrs::auth::cert::{self};
+use vaultrs::client::{Client, VaultClient, VaultClientSettingsBuilder};
+use vaultrs::error::ClientError;
+use vaultrs::sys::auth;
+
+use crate::common::{PORT, VERSION};
+
+#[cfg(not(feature = "native-tls"))]
```
Comment (Haennetz) [inline tests/cert.rs:22]:
```suggestion
```

File: tests/cert.rs [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+        let client = VaultClient::new(
+            VaultClientSettingsBuilder::default()
+                .address(format!("https://localhost:{PORT}"))
+                .token(server.token.clone())
+                .ca_certs(vec![ca_cert_path])
+                .identity(Some(identity))
+                .build()
+                .unwrap(),
+        )
+        .unwrap();
```
Comment (Haennetz) [inline tests/cert.rs:60]: This Handles the error if it is the expected `incompatible TLS identity type` and if the feature `native-tls` true and returns the function.
```suggestion
let client = match VaultClient::new(
            VaultClientSettingsBuilder::default()
                .address(format!("https://localhost:{PORT}"))
                .token(server.token.clone())
                .ca_certs(vec![ca_cert_path])
                .identity(Some(identity))
                .build()
                .unwrap(),
        ) {
            Ok(c) => c,
            Err(err) => {
                assert!(err
                    .source()
                    .unwrap()
                    .source()
                    .unwrap()
                    .to_string()
                    .eq("incompatible TLS identity type"));
                assert!(cfg!(feature = "native-tls").eq(&true));
                return;
            }
        };
```

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -87,13 +87,20 @@ jobs:
...
         with:
           command: test
           args: --all-features
+      - name: Run cargo test --features rustls
+        uses: actions-rs/cargo@v1
+        env:
+          RUST_TEST_THREADS: 1
+        with:
+          command: test
+          args: --features rustls
```
Comment (Haennetz) [inline .github/workflows/ci.yml:103]: This uses the default feature which includes the `rustls` 
```suggestion
      - name: Run cargo test
        uses: actions-rs/cargo@v1
        env:
          RUST_TEST_THREADS: 1
        with:
          command: test
```

**Review by Haennetz — COMMENTED** [review]
With the suggestions we should get rid of the warnings.

**Comment by avandecreme** [comment]
Suggestion applied

**Comment by Haennetz** [comment]
Thanks :+1:

**Pull request merged**